TrackletAlgorithm/TrackBuilder.h

```diff
@@ -0,0 +1,276 @@
+    ap_uint<1> barrel_valid[NFMBarrel];
+    ap_uint<1> disk_valid[NFMDisk];
+#pragma HLS array_partition variable=barrel_index complete dim=0
+#pragma HLS array_partition variable=disk_index complete dim=0
+#pragma HLS array_partition variable=barrel_valid complete dim=0
+#pragma HLS array_partition variable=disk_valid complete dim=0
+
+    // First determine the minimum tracklet ID from the current set of full
+    // matches.
+    barrel_min_id : for (unsigned short j = 0; j < NFMBarrel; j++) {
+
+      const auto &barrel_stub_0 = barrel_fm[j][barrel_read_index[j]];
+      const auto &barrel_id_0 = barrel_stub_0.id();
+      barrel_index[j] = barrel_stub_0.index();
+      barrel_valid[j] = (!empty && barrel_id_0 != kInvalidTrackletID);
+
+      // Compare the given barrel and disk IDs against each barrel ID.
+      barrel_barrel_id_comp : for (unsigned short k = 0; k < NFMBarrel; k++) {
+        const auto &barrel_stub_1 = barrel_fm[k][barrel_read_index[k]];
+        const auto &barrel_id_1 = barrel_stub_1.id();
+        barrel_valid[j] = (barrel_valid[j] && barrel_id_0 <= barrel_id_1);
+      }
+      // Compare the given barrel and disk IDs against each disk ID.
+      barrel_disk_id_comp : for (unsigned short k = 0; k < NFMDisk; k++) {
+        const auto &disk_stub_1 = disk_fm[k][disk_read_index[k]];
+        const auto &disk_id_1 = disk_stub_1.id();
+        barrel_valid[j] = (barrel_valid[j] && barrel_id_0 <= disk_id_1);
+      }
+
+      min_id = (barrel_valid[j] ? barrel_id_0 : min_id);
+    }
+
+    disk_min_id : for (unsigned short j = 0; j < NFMDisk; j++) {
+
+      const auto &disk_stub_0 = disk_fm[j][disk_read_index[j]];
+      const auto &disk_id_0 = disk_stub_0.id();
+      disk_index[j] = disk_stub_0.index();
+      disk_valid[j] = (!empty && disk_id_0 != kInvalidTrackletID);
+
+      // Compare the given barrel and disk IDs against each barrel ID.
+      disk_barrel_id_comp : for (unsigned short k = 0; k < NFMBarrel; k++) {
+        const auto &barrel_stub_1 = barrel_fm[k][barrel_read_index[k]];
+        const auto &barrel_id_1 = barrel_stub_1.id();
+        disk_valid[j] = (disk_valid[j] && disk_id_0 <= barrel_id_1);
+      }
+      // Compare the given barrel and disk IDs against each disk ID.
+      disk_disk_id_comp : for (unsigned short k = 0; k < NFMDisk; k++) {
+        const auto &disk_stub_1 = disk_fm[k][disk_read_index[k]];
+        const auto &disk_id_1 = disk_stub_1.id();
+        disk_valid[j] = (disk_valid[j] && disk_id_0 <= disk_id_1);
+      }
+
+      min_id = (disk_valid[j] ? disk_id_0 : min_id);
+    }
+
+    // Initialize a TrackFit object using the tracklet parameters associated
+    // with the minimum tracklet ID.
+    const TCIDType &TCID = (min_id != kInvalidTrackletID) ? (min_id >> 7) : TrackletIDType(0);
+    TrackFit track(nTracks, TCID >> 4);
+    const IndexType &trackletIndex = (min_id != kInvalidTrackletID) ? (min_id & TrackletIDType(0x7F)) : TrackletIDType(0);
```

Comment:
Eliminate magic number "4" in "TCID >> 4"

Reply:
Done.